internal/handlers/pg/pgdb/settings.go

```diff
@@ -0,0 +1,223 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pgdb
+
+import (
+	"context"
+	"fmt"
+	"hash/fnv"
+
+	"github.com/jackc/pgx/v4"
+	"golang.org/x/exp/slices"
+
+	"github.com/FerretDB/FerretDB/internal/fjson"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/util/must"
+)
+
+const (
+	// Internal collections prefix.
+	collectionPrefix = "_ferretdb_"
+
+	// PostgreSQL max table name length.
+	maxTableNameLength = 63
+)
+
+// CreateSettingsTable creates FerretDB settings table if it doesn't exist.
+func (pgPool *Pool) CreateSettingsTable(ctx context.Context, db string) error {
+	tables, err := pgPool.tables(ctx, db)
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	if slices.Contains(tables, collectionPrefix+"settings") {
+		return ErrAlreadyExist
+	}
+
+	sql := `CREATE TABLE ` + pgx.Identifier{db, collectionPrefix + "settings"}.Sanitize() + ` (settings jsonb)`
+	_, err = pgPool.Exec(ctx, sql)
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	settings := must.NotFail(types.NewDocument("collections", must.NotFail(types.NewDocument())))
+	sql = fmt.Sprintf(`INSERT INTO %s (settings) VALUES ($1)`, pgx.Identifier{db, collectionPrefix + "settings"}.Sanitize())
+	_, err = pgPool.Exec(ctx, sql, must.NotFail(fjson.Marshal(settings)))
+	if err != nil {
+		return lazyerrors.Error(err)
+	}
+
+	return nil
+}
+
+// GetTableName returns the name of the table for given collection or error.
+func (pgPool *Pool) GetTableName(ctx context.Context, db, collection string) (string, error) {
+	var err error
+
+	schemaExists, err := pgPool.schemaExists(ctx, db)
+	if err != nil {
+		return "", lazyerrors.Error(err)
+	}
+
+	if !schemaExists {
+		return GetTableNameFormatted(collection), nil
+	}
+
+	var tables []string
+	if tables, err = pgPool.tables(ctx, db); err != nil {
+		return "", lazyerrors.Error(err)
+	}
+	if !slices.Contains(tables, collectionPrefix+"settings") {
+		err = pgPool.CreateSettingsTable(ctx, db)
+		if err != nil {
+			return "", lazyerrors.Error(err)
+		}
+	}
+
+	tx, err := pgPool.Begin(ctx)
+	if err != nil {
+		return "", lazyerrors.Error(err)
+	}
+	defer func() {
+		if err != nil {
+			must.NoError(tx.Rollback(ctx))
+			return
+		}
+		must.NoError(tx.Commit(ctx))
+	}()
+
+	settings, err := pgPool.getSettingsTable(ctx, tx, db)
+	if err != nil {
+		return "", lazyerrors.Error(err)
+	}
+
+	collectionsDoc := must.NotFail(settings.Get("collections"))
+	collections, ok := collectionsDoc.(*types.Document)
+	if !ok {
+		return "", lazyerrors.Errorf("invalid collections document: %v", collectionsDoc)
```

Comment:
Let's create meaningful error messages: 
* it's not a document, but why it should be a document?

Reply:
Fixed